### PR TITLE
runtime: fix nil C string handling

### LINKED ISSUE #889: Feature Request: implemnts some methods to build a basic
I just want to compile a demo base on https://github.com/mattn/go-sqlite3 that used cgo, all cgo compiling problems are solved in (https://github.com/goplus/llgo/pull/881), but there's some methods not implemented.

Methods required by https://github.com/mattn/go-sqlite3 :
```go
reflect.Copy
reflect.Indirect
reflect.Value.Convert
reflect.Value.OverflowFloat
reflect.Value.OverflowInt
reflect.Value.OverflowUint
runtime.Stack
sync.(*Mutex).lockSlow
sync.(*Mutex).unlockSlow
sync.(*RWMutex).RLocker
time.LoadLocation
time.Time.In
```

And some other methods needed:

```go
runtime.FuncForPC
runtime.(*Func).Name
```

I would be very grateful if someone could implement these.

Sqlite3 demo:

```go
package main

import (
	"database/sql"
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"log"
	"os"

	_ "github.com/mattn/go-sqlite3"
)

type Tag struct {
	Name    string `json:"name"`
	Country string `json:"country"`
}

func (t *Tag) Scan(value interface{}) error {
	return json.Unmarshal([]byte(value.(string)), t)
}

func (t *Tag) Value() (driver.Value, error) {
	b, err := json.Marshal(t)
	return string(b), err
}

func main() {
	os.Remove("./foo.db")

	db, err := sql.Open("sqlite3", "./foo.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	_, err = db.Exec(`create table foo (tag jsonb)`)
	if err != nil {
		log.Fatal(err)
	}

	stmt, err := db.Prepare("insert into foo(tag) values(?)")
	if err != nil {
		log.Fatal(err)
	}
	defer stmt.Close()
	_, err = stmt.Exec(`{"name": "mattn", "country": "japan"}`)
	if err != nil {
		log.Fatal(err)
	}
	_, err = stmt.Exec(`{"name": "michael", "country": "usa"}`)
	if err != nil {
		log.Fatal(err)
	}

	var country string
	err = db.QueryRow("select tag->>'country' from foo where tag->>'name' = 'mattn'").Scan(&country)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(country)

	var tag Tag
	err = db.QueryRow("select tag from foo where tag->>'name' = 'mattn'").Scan(&tag)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println(tag.Name)

	tag.Country = "日本"
	_, err = db.Exec(`update foo set tag = ? where tag->>'name' == 'mattn'`, &tag)
	if err != nil {
		log.Fatal(err)
	}

	err = db.QueryRow("select tag->>'country' from foo where tag->>'name' = 'mattn'").Scan(&country)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(country)
}
```

### PR DESCRIPTION
**Pull request by cpunion**
Fixes #889.

## Root cause

`github.com/mattn/go-sqlite3` uses both of these C-string conversion paths:

- cgo `C.GoString` / `C.GoStringN` -> `runtime.GoString` / `runtime.GoStringN`
- `github.com/goplus/lib/c.GoString` -> `llgo.string` -> `runtime.StringFromCStr`

For expressions such as `select 42`, SQLite may return a null `decltype` pointer.
`go-sqlite3` calls `C.GoString` on that pointer. gc treats `GoString(nil)` as `""`, but llgo was calling `strlen(NULL)` and crashing.

## What changed

- Make `runtime.GoString` / `runtime.GoStringN` return an empty string for nil pointers.
- Make `runtime.StringFromCStr` return an empty string for nil pointers.
- Add an llgo regression test for nil C strings.

## Verification

- `go test ./test -run TestCGoStringNil -count=1`
- `go build -tags=dev -o /tmp/llgo-889 ./cmd/llgo`
- `/tmp/llgo-889 test ./test -run TestCGoStringNil -count=1`
- `/tmp/llgo-889 test ./test/std/database/sql -count=1`
- `/tmp/llgo-889 run ./cl/_testgo/cgobasic`
- minimal external module using `github.com/mattn/go-sqlite3` and `db.QueryRow("select 42").Scan(&n)`
- full sample from #889 now prints:
  - `japan`
  - `mattn`
  - `日本`

### CODE REVIEW

File: test/cgo_string_test.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+func TestCGoStringNil(t *testing.T) {
+	var p *c.Char
+	if got := c.GoString(p); got != "" {
+		t.Fatalf("GoString(nil) = %q, want empty string", got)
+	}
+	if got := c.GoString(p, 0); got != "" {
+		t.Fatalf("GoStringN(nil, 0) = %q, want empty string", got)
+	}
```
[the remark this answers is not in the document]
Comment (cpunion) [reply test/cgo_string_test.go:19]: I tightened `GoStringN` in the follow-up commit to match the standard runtime more closely: only `n <= 0` returns `""` here. The sqlite fix depends on `GoString(nil) -> ""`, not on `GoStringN(nil, positive)` being accepted, so I am intentionally not extending the test in that direction.

File: test/cgo_string_test.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+	"github.com/goplus/lib/c"
+)
+
+func TestCGoStringNil(t *testing.T) {
+	var p *c.Char
+	if got := c.GoString(p); got != "" {
+		t.Fatalf("GoString(nil) = %q, want empty string", got)
+	}
+	if got := c.GoString(p, 0); got != "" {
```
[the remark this answers is not in the document]
Comment (cpunion) [reply test/cgo_string_test.go:17]: I intentionally did not add a `GoStringN(nil, positive)` assertion in the follow-up. The sqlite crash is fixed by `GoString(nil) -> ""`, and I tightened `GoStringN` to stay closer to the standard runtime rather than continuing to accept `nil` with a positive length. `StringFromCStr(nil)` is also guarded now, but I did not add a separate test for it because it is a runtime-internal helper and there is no clean existing `go test` entry point for it in this repo without adding linkname-only test plumbing.

File: runtime/internal/runtime/z_cgo.go [hunk 3]
```diff
@@ -34,10 +34,16 @@ func CBytes(b []byte) *int8 {
...
+	}
 	return GoStringN(p, int(c.Strlen(p)))
 }
 
 func GoStringN(p *int8, n int) string {
+	if p == nil || n <= 0 {
+		return ""
+	}
 	return string((*[1 << 30]byte)(unsafe.Pointer(p))[:n:n])
 }
```
[the remark this answers is not in the document]
Comment (cpunion) [reply runtime/internal/runtime/z_cgo.go:49]: I checked `cmd/cgo` and `runtime.gobytes` in the standard runtime before changing this. `C.GoBytes` maps to `runtime.gobytes`, and `gobytes` does not special-case `nil && n > 0`; it only handles `n == 0` and range checks before `memmove`. I am leaving `GoBytes` unchanged here to stay aligned with gc semantics.

**Review by visualfc — APPROVED** [review]
